glyphs-reader/src/font.rs

```diff
@@ -264,3 +294,296 @@ impl Path {
+                            v3_axes.push(Axis {
+                                name: v2_axis.get("Name").unwrap().as_str().unwrap().into(),
+                                tag: tag.into(),
+                            });
+                        }
+                    }
+                }
+            }
+            axes_idx.map(|idx| custom_params.remove(idx));
+
+            if custom_params.is_empty() {
+                self.other_stuff.remove("customParameters");
+            }            
+        }
+        Ok(tags)
+    }
+
+    fn v2_to_v3_metrics(&mut self) -> Result<(), Error> {
+        // metrics are in parallel arrays in v3
+        let v3_metric_names = ["ascender".to_string(), "baseline".to_string(), "descender".to_string(), "cap height".to_string(), "x-height".to_string()];
+        let v2_metric_names = ["ascender".to_string(), "baseline".to_string(), "descender".to_string(), "capHeight".to_string(), "xHeight".to_string()];
+        assert!(v2_metric_names.len() == v3_metric_names.len());
+
+        // setup root storage for the basic metrics
+        let mut metrics = Vec::new();
+        for name in v3_metric_names {
+            metrics.push(Plist::Dictionary(BTreeMap::from([("type".to_string(), Plist::String(name.clone()))])));
+        }        
+        self.other_stuff.insert("metrics".into(), Plist::Array(metrics));
+
+        // in each font master setup the parallel array
+        for master in self.font_master.iter_mut() {
+            let mut metric_values = Vec::new();
+            for v2_name in v2_metric_names.iter() {
+                let mut dict = BTreeMap::new();
+                if let Some(Plist::Integer(value)) = master.other_stuff.remove(v2_name) {
+                    // leave blank for 0
+                    if value != 0 {
+                        dict.insert("pos".to_string(), Plist::Integer(value));
+                    }                    
+                }
+                metric_values.push(Plist::Dictionary(dict));
+            }
+            master.other_stuff.insert("metricValues".into(), Plist::Array(metric_values));
+        }
+        Ok(())
+    }
+
+    fn v2_to_v3_shapes(&mut self) -> Result<(), Error> {
+        // shapes in v3 encompasses paths and components in v2
+        for glyph in self.glyphs.iter_mut() {
+            // v2 uses single codepoint strings, turn into int to match v3 for now
+            // In time we will likely parse unicode more carefully
+            glyph.other_stuff.entry("unicode".into()).and_modify(|v| {
+                if let Plist::String(val) = v {
+                    *v = Plist::Integer(i64::from_str_radix(val, 16).unwrap())
+                }
+            });
+
+            // Paths and components combine in shapes
+            // TODO components
+            for layer in glyph.layers.iter_mut() {
+                let paths = layer.paths.take();
+                if paths.is_some() {
+                    match &mut layer.shapes {
```

Comment:
I believe if it has `paths` (v2 only) is doesn't have `shapes` (v3 only) and viceversa

Reply:
I think that's true but this way felt more correct despite arguably being overkill